Create orgs event button

theJenix opened the pull request:
This pull request is to add a "Create an event" page directly to an org's page, and repopulate the combo box in the create events page with the name of the aforementioned org.

theJenix committed:
first cut at markdown preview functionality

added preview button on edit orgs page
added /orgs/preview url (in index.js), handler code in orgs.js,
preview.jade to render markdown and code in create-org.jade to send
description to server, and load results in a preview pane below the
buttons.

theJenix committed:
create events from org page

added "create an event" button on the org page
added code to make sure that the correct org is preselected in the
"create event" page

theJenix committed:
Revert "first cut at markdown preview functionality"

This reverts commit 3f464b93fe4e4f26ba047d31051e20fde3bb3bdd.

theJenix committed:
safeguarded selected org param processing

moved core to look up selected org from the http param back into .js
file (instead of relying on the .jade file).  this is a bit safer, as
we can assume here all of the risks that comes from pulling from a
query param, but in the jade file those risks are obscured

theJenix commented:
You can ignore the "preview functionality" commits...that's something else I'm working on, and I didn't know how to use Git effectively (hence the revert).

thelastnode commented on routes/events.js:
```diff
@@ -74,9 +74,21 @@ exports.create = function(req, res, next) {
         }
+        //NOTE: look up the org in the query here, rather than pass it to the jade file
+        //...slightly more secure because we know the origin of the data here (http query) and can
+        // take appropriate protective steps, rather than relying on the .jade file to be safe
```
The .jade file is rendered on the server side, right below where it says res.render, the security is the same.

theJenix replied:
Yea, in retrospect, this comment is probably not as good as it could be.  What I meant to say was, it's obvious here where that name comes from (the query), whereas in the .jade file, selOrg comes from whatever js is rendering it.  In this case, it's probably not a big deal, but in a more sensitive situation passing the query on to the .jade file would require anyone modifying the .jade file to validate and safeguard the parameter and I thought it would be better to deal with all that in the .js file.

thelastnode commented on routes/events.js:
```diff
@@ -74,9 +74,21 @@ exports.create = function(req, res, next) {
+        //...slightly more secure because we know the origin of the data here (http query) and can
+        // take appropriate protective steps, rather than relying on the .jade file to be safe
+        var selOrg;
+        for (inx in orgs) {
```
This loops over each org in orgs, not each index in orgs.

theJenix replied:
Actually, in javascript the for-each syntax gives you the index in the loop variable, not the actual object.  This does work to iterate thru the indexes in orgs.

thelastnode replied:
Oh, you're right, my mistake. As a style note, for non-member variables, prefer `variable_name` to `variableName` (I'm going to make an issue to make a style guide)

theJenix replied:
no worries...and ok, no problem...do you want me to fix it?  or will you fix it when you pull the changes in?

thelastnode commented on routes/events.js:
```diff
@@ -74,9 +74,21 @@ exports.create = function(req, res, next) {
+        // take appropriate protective steps, rather than relying on the .jade file to be safe
+        var selOrg;
+        for (inx in orgs) {
+            var org = orgs[inx];
```
This only works because it returns undefined, so none of them are matched. You likely are only the admin for one org, so it defaults to showing that one since none have the selected flag.

theJenix replied:
I tested with 2 orgs on my local machine, and it correctly selected between the 2 (based on the query parameter).  See the comment about javascript for-each loops above.

thelastnode commented on routes/events.js:
```diff
@@ -74,9 +74,21 @@ exports.create = function(req, res, next) {
+                selOrg = org.name;
+                break;
+            }
+        }
```
Let's move this logic into the template, since it's presentation logic, makes it a pretty two-line diff too :)

```
option(value=org.id, selected=(req.query.org == org.name))= org.name
```

theJenix replied:
Oh, you can access the req directly in the .jade file?  Yea, that's better.  Let me make that change and give it a whirl.

theJenix committed:
reorganized selected org functionality

moved "selected org" functionality entirely into .jade file, and
removed for loop from events.js (much cleaner this way)

theJenix commented:
Don't know if you got the notification, but I committed the change to move the selected org check to the .jade file.  Incidentally, removing that for loop also removed the variable name issue as well.  Two birds, one stone.
